File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.js
function solution(s) {
  var answer = 0;

  let prev = 0, cur = 0, first;
  [...s].map((str, idx) => {
    if (!first) {
      first = str;
      prev++;
    } else if (first === str) {
      prev++;
    } else {
      cur++;
    }

    if (prev === cur) {
      prev = 0;
      cur = 0;
      first = undefined;
      answer++;
    }
  })

  if (prev !== 0 || cur !== 0) {
    answer++;
  }

  console.log(answer);
  return answer;

}

solution("banana");
// solution("abracadabra");
// solution("aaabbaccccabba");

File: 프로그래머스/1/140108. 문자열 나누기/문자열 나누기.py
def solution(s: str) -> int:
    answer = 0
    first = None
    same = 0
    diff = 0
    for ch in s:
        if first is None:
            first = ch
            same = 1
            diff = 0
            continue
        if ch == first:
            same += 1
        else:
            diff += 1
        if same == diff:
            answer += 1
            first = None
            same = 0
            diff = 0
    if first is not None:
        answer += 1
    return answer
